### ISSUE TITLE
Edge behavior in `jax.scipy.special.betainc`

### ISSUE THREAD

**Issue by mdhaber** [jax-ml/jax#21900]
### Description

`jax.scipy.special.betainc` seems to have trouble with very small values of the parameter `a`, at least for certain values of `b` and `x`.

```python3
import matplotlib.pyplot as plt
import numpy as np
from scipy.special import betainc as betainc_scipy
import jax.numpy as xp
from jax.scipy.special import betainc as betainc_jax

a = np.logspace(-40, -1, 300)
b = 1
x = 0.25
plt.loglog(a, betainc_scipy(a, b, x), label='scipy')
plt.loglog(a, betainc_jax(xp.asarray(a), b, x), label='jax')
plt.xlabel('a')
plt.ylabel('betainc(a, 1, 0.25)')
plt.legend()
```
<img width="613" alt="image" src="https://github.com/scipy/scipy/assets/6570539/b3c6a70d-1b5d-496d-b4ef-93a266b46bf1">

I know that it is difficult to guarantee accuracy to machine precision for all possible combinations of input : ) Just thought I'd point out this problem spot since it came up in SciPy testing (scipy/scipy#20963).

---

There is a separate issue related to edge case behavior in `betainc` that I thought I should bring up: there are some edge cases where `betainc` returns NaN but another result would be more useful (e.g. in Harrell-Davis quantile estimates). To some extent, the "correct" result is a matter of interpretation, but SciPy recently addressed similar cases in scipy/scipy#22425. While we're waiting for an Array API special function extension (data-apis/array-api#725), please consider adopting this behavior.

<details>

```py
import jax.numpy as np
from jax.scipy import special
cases = [((0.0, 0.0, 0.0), np.nan),
         ((0.0, 0.0, 0.5), np.nan),
         ((0.0, 0.0, 1.0), np.nan),
         ((np.inf, np.inf, 0.0), np.nan),
         ((np.inf, np.inf, 0.5), np.nan),
         ((np.inf, np.inf, 1.0), np.nan),
         ((0.0, 1.0, 0.0), 0.0),
         ((0.0, 1.0, 0.5), 1.0),
         ((0.0, 1.0, 1.0), 1.0),
         ((1.0, 0.0, 0.0), 0.0),
         ((1.0, 0.0, 0.5), 0.0),
         ((1.0, 0.0, 1.0), 1.0),
         ((0.0, np.inf, 0.0), 0.0),
         ((0.0, np.inf, 0.5), 1.0),
         ((0.0, np.inf, 1.0), 1.0),
         ((np.inf, 0.0, 0.0), 0.0),
         ((np.inf, 0.0, 0.5), 0.0),
         ((np.inf, 0.0, 1.0), 1.0),
         ((1.0, np.inf, 0.0), 0.0),
         ((1.0, np.inf, 0.5), 1.0),
         ((1.0, np.inf, 1.0), 1.0),
         ((np.inf, 1.0, 0.0), 0.0),
         ((np.inf, 1.0, 0.5), 0.0),
         ((np.inf, 1.0, 1.0), 1.0)]

for args, ref in cases:
    print(special.betainc(*args), ref)
```

Produces: 
(actual, desired)
```
nan nan
nan nan
nan nan
nan nan
nan nan
nan nan
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
```

</details>

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.25.2
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='e901fac133dc', release='6.1.85+', version='#1 SMP PREEMPT_DYNAMIC Sun Apr 28 14:29:16 UTC 2024', machine='x86_64')

**Comment by rajasekharporeddy** [comment]
Hi @mdhaber 

JAX typically uses single-precision floating-point numbers, while SciPY defaults to double precision. This difference in precision can lead to slightly different results, especially when working with very small numbers. If the double precision is enabled in JAX, then JAX yields the results that are consistent with SciPy even with very small numbers:

```python
import jax

jax.config.update('jax_enable_x64', True)

import matplotlib.pyplot as plt
import numpy as np
from scipy.special import betainc as betainc_scipy
import jax.numpy as xp
from jax.scipy.special import betainc as betainc_jax

a = np.logspace(-40, -1, 300)
b = 1
x = 0.25
plt.loglog(a, betainc_scipy(a, b, x), label='scipy')
plt.loglog(a, betainc_jax(xp.asarray(a), b, x), label='jax')
plt.xlabel('a')
plt.ylabel('betainc(a, 1, 0.25)')
plt.legend()
``` 
![image](https://github.com/google/jax/assets/153593711/96a09748-4da8-4e2c-82b9-248953187350)

Please find the [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/598a3964249675e82ceaefe17367a9ce/-21900.ipynb) for reference.

Thank you.

**Comment by mdhaber** [comment]
Thanks!

**Comment by rajasekharporeddy** [comment]
Hi @mdhaber 

According to [scipy/scipy/#8495 (Comment)](https://github.com/scipy/scipy/issues/8495#issuecomment-370788812), SciPy do all the internal (low level c) calculations in `float64` even if the input is `float32` or other. But JAX do it in `float32` itself. That is why this difference occurs.

**Comment by pearu** [comment]
Whatever are the reasons for scipy to use float64 internally (one practical reason could be that there are no float32 implementation available for scipy, for instance), evaluating functions using float32 correctly requires the usage of an algorithm that can properly handle overflows, underflows, or cancellations. Using higher precision is a typical cheap trick to avoid paying attention to these fp errors in implementations of the function algorithms to keep algorithms simple.
So, I wonder what is the location of jax.scipy.special.betainc implementation which may provide explanations for the behavior observed in this issue.

**Comment by jakevdp** [comment]
JAX's implementation is here: https://github.com/google/jax/blob/2b728d55b6054bba8ae26b3523722e80d660e771/jax/_src/lax/special.py#L182-L190

**Comment by mdhaber** [comment]
> SciPy do all the internal (low level c) calculations in float64 even if the input is float32 or other.

But that comment is about `scipy.ndimage.affine_transform`, not `scipy.special.betainc`.

I confirmed with @steppi that SciPy now uses Boost's `ibeta` for `betainc`, and the types seem to be preserved in the calculation.

Here is where `betainc` is defined in terms of `ibeta`.
https://github.com/scipy/scipy/blob/e36e728081475466d2faae65e1dfecfa2314c857/scipy/special/functions.json#L118-L123

Here is where `ibeta` is used for `float` and `double` instantiations of the function.
https://github.com/scipy/scipy/blob/e36e728081475466d2faae65e1dfecfa2314c857/scipy/special/boost_special_functions.h#L194-L204

and Boost's `ibeta` is templated:
https://beta.boost.org/doc/libs/1_68_0/libs/math/doc/html/math_toolkit/sf_beta/ibeta_function.html

**Comment by vfdev-5** [comment]
Checking the same code on GPU, we have a bit different plots:
```python
import matplotlib.pyplot as plt
import numpy as np
from scipy.special import betainc as betainc_scipy
import jax.numpy as xp
from jax.scipy.special import betainc as betainc_jax

a = np.logspace(-40, -1, 300)
b = 1
x = 0.25

output = betainc_jax(xp.asarray(a), b, x)

plt.loglog(a, betainc_scipy(a, b, x), label='scipy')
plt.loglog(a, output, label='jax')
plt.xlabel('a')
plt.ylabel('betainc(a, 1, 0.25)')
plt.legend()

print(output.devices(), output.dtype)
# {cuda(id=0)} float32
```
![image](https://github.com/google/jax/assets/2459423/030daddb-102c-42c5-9c4f-21e7d5a74652)

and 
```python
import matplotlib.pyplot as plt
import numpy as np
from scipy.special import betainc as betainc_scipy
import jax.numpy as xp
from jax.scipy.special import betainc as betainc_jax

a0 = np.finfo(np.float32).smallest_normal
b = np.float32(1.)
x = np.float32(0.25)
factor = np.float32(10)
a = np.logspace(np.log10(a0), np.log10(a0*factor), 300, dtype=np.float32)

output = betainc_jax(xp.asarray(a), b, x)
plt.loglog(a, betainc_scipy(a, b, x), label='scipy')
plt.loglog(a, output, label='jax')
plt.xlabel('a')
plt.ylabel('betainc(a, 1, 0.25)')
plt.legend()

print(output.devices(), output.dtype)
# {cuda(id=0)} float32
```
![image](https://github.com/google/jax/assets/2459423/2b49a11c-35cd-4754-9bcf-1ec216ca10c2)

**Comment by mdhaber** [comment]
Thanks for the investigation @vfdev-5. I extended the top post with a description of a separate issue regarding edge case behavior, which is probably easier to address.

**Comment by pearu** [comment]
Heads up: https://github.com/jax-ml/jax/pull/27107 fixes all issues reported here.

**Comment by mdhaber** [comment]
Thank you!
